### PR TITLE
Run imports in CI

### PR DESCRIPTION
**Pull request by GeoWill**
The aim here is to run import scripts as they get merged in CI 
Closes https://github.com/DemocracyClub/UK-Polling-Stations/issues/4724

## Trello
https://trello.com/c/0bSKpmSv

## Run it locally
`./manage.py run_new_imports -f 4078f6ac -t fcf57110a`

Thoughts on the approach in #x-will. Key point is I made it a management command because it was easier to pinch code from other bits, and seemed better organised and easier to test to me.

```[tasklist]
### PR Checklist
<!-- Not all the items in this list will be relevant for every repo -->
<!-- When creating a Pull Request please delete items as necessary, leaving those that are relevant.  -->

Check off items once the PR addresses them.

- [ ] References a specific issue or if not, describes the bug or feature in detail
- [ ] Note what card in Trello this work relates to
- [ ] Instructions for how reviewers can test the code locally
- [x] Tests have been added and/or updated
- [ ] Did I use the clear and concise names for variables and functions?
- [ ] Did I explain all possible solutions and why I chose the one I did?
- [ ] Have I rebased with the latest version of master?
- [ ] Added PR labels
```

### CODE REVIEW

**Comment by GeoWill** [comment]
I seem to have broken coveralls :roll_eyes:

File: polling_stations/apps/data_importers/base_importers.py [hunk 1]
```diff
@@ -194,6 +204,23 @@ def get_base_folder_path(self):
...
+        """
+        if not hasattr(self, "elections"):
+            return False
+        for date_str in self.elections:
+            try:
+                election_date = datetime.date.fromisoformat(date_str)
+                if election_date >= datetime.date.today():
+                    return True
+            except ValueError:
+                pass
```
Comment (symroe) [inline polling_stations/apps/data_importers/base_importers.py:221]: Is it worth raising a warning here?

File: polling_stations/apps/data_importers/management/commands/run_new_imports.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+            try:
+                call_command(script_path.stem, **opts)
+                self.summary.append(
+                    (
+                        "INFO",
+                        f"Ran import script: {script_path.name}",
+                    )
+                )
+            except:
+                # usually we want to handle a specific exception, but in this situation
```
Comment (symroe) [inline polling_stations/apps/data_importers/management/commands/run_new_imports.py:98]: Is this true? Why don't you want to fail CI if the import script can't run?

Comment (symroe) [reply polling_stations/apps/data_importers/management/commands/run_new_imports.py:98]: Also, it's a good idea to at least catch `Exception`. This means that other

File: polling_stations/apps/data_importers/management/commands/run_new_imports.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            "include_past_elections": False,
+        }
+
+        sys.stdout.write(
+            f"Comparing repo between {from_sha} and {to_sha}\n"
+            f"\tFrom: https://github.com/DemocracyClub/UK-Polling-Stations/commit/{git_rev_parse(from_sha)}\n"
+            f"\tTo: https://github.com/DemocracyClub/UK-Polling-Stations/commit/{git_rev_parse(from_sha)}\n"
+        )
+
+        # __import__("ipdb").set_trace()
```
Comment (symroe) [inline polling_stations/apps/data_importers/management/commands/run_new_imports.py]: Clean up

File: polling_stations/apps/data_importers/management/commands/run_new_imports.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        if has_imports and not has_application:
+            sys.stdout.write("Only import scripts have changed\n")
```
Comment (symroe) [inline polling_stations/apps/data_importers/management/commands/run_new_imports.py]: I think you want `self.stdout` in these. `sys.stdout.write` is the same as `print`, but it's useful to use whatever `self.stdout` is set to in management commands

File: polling_stations/apps/data_importers/tests/test_run_new_imports.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        "polling_stations/apps/data_importers/management/commands/import_barrow_in_furness.py",
+        "polling_stations/apps/data_importers/management/commands/import_eoni.py",
+        "polling_stations/apps/data_importers/management/commands/import_wigan.py",
+        "polling_stations/settings/constants/councils.py",
+        "polling_stations/templates/base.html",
+        "polling_stations/templates/postcode_view.html",
+        "requirements/base.txt",
+    ]
+    assert expected_empty == get_paths_changed(
+        "1012c398b13d2e9c87f718b87e07ee9cd1c26222",
```
Comment (symroe) [inline polling_stations/apps/data_importers/tests/test_run_new_imports.py:52]: I see what you've done here, but something feels…odd about this code checking it's own git history. I wonder what happens if, for whatever reason, we change the hashed (for example, what if we pull import scripts into a new repo? Maybe this PR is moot in that case, but you get my point).

To mitigate changes, can you leave a comment as to how you found these magic values?

File: .circleci/config.yml [hunk 6]
```diff
@@ -233,6 +273,18 @@ workflows:
         min-size: 1
         max-size: 2
         desired-capacity: 1
+    - post_deploy_tasks:
+        name: "Development: Post-Deploy Tasks"
+        requires:
+        - "Development: AWS CodeDeploy"
+        context: [CircleCIAPIUser, deployment-development-wdiv]
+        filters: { branches: { only: [ main, master ] } }
+    - pre_deploy_tasks:
```
Comment (symroe) [inline .circleci/config.yml:282]: Would you mind ordering these in the way the tasks are ordered? Like, move `pre` before all other steps, and `post` after the others?

**Review by symroe — COMMENTED** [review]
I think you need to implement `--exit_early` and call this on all jobs, otherwise the deploy will always happen?

File: polling_stations/apps/data_importers/management/commands/run_new_imports.py (hunk 4 above)
Comment (GeoWill) [reply polling_stations/apps/data_importers/management/commands/run_new_imports.py]: rebase fail - I did that here 83ccb291a637243967da56b12e27ce37f7e4a667

File: polling_stations/apps/data_importers/management/commands/run_new_imports.py (hunk 3 above)
Comment (GeoWill) [reply polling_stations/apps/data_importers/management/commands/run_new_imports.py]: ditto.

File: polling_stations/apps/data_importers/base_importers.py (hunk 1 above)
Comment (GeoWill) [reply polling_stations/apps/data_importers/base_importers.py:221]: Good shout.

File: polling_stations/apps/data_importers/management/commands/run_new_imports.py (hunk 2 above)
Comment (GeoWill) [reply polling_stations/apps/data_importers/management/commands/run_new_imports.py:98]: Right, so my thinking here was that we don't want failed import scripts to block deploys. But I think I actually agree with you. Most of our import script changes are single commits, so it's simple enough to just drop that commit if there's an issue. I think I might go with your recommendation of catching `Exception` as a starting point, then refactor further to take it out completely.

**Comment by GeoWill** [comment]
> I think you need to implement --exit_early and call this on all jobs, otherwise the deploy will always happen?

So I haven't implemented that because I didn't think it was needed. 

### Only import scripts
- Command is called as in the `pre-deploy` job. This:
  - runs the import scripts
  - sets the ssm parameter
  - quits the circle jobs

### Only app
- No scripts are run
  - Checking time seems negligible

### Mix of the two
- pre-deploy
  - Scripts don't run because --post-deploy flag isn't set
- post-deploy
  - scripts run, but nothing to exit from.

**Comment by symroe** [comment]
> quits the circle jobs

`inigo-montoya.gif`.

So, `args = ["circleci-agent", "step", "halt"]` doesn't stop the _workflow_, it stops the _step_ cleanly. However the workflow continues, meaning that, as things stand, a deploy will happen each time.

[This](https://github.com/DemocracyClub/circle-ci-conditional-job-test/blob/new-new-stroud/magic.py#L54-L62) was my attempt at stopping the workflow, but that was the approach that made it look like all the jobs had failed, so we chose to [add a check to each job](https://github.com/DemocracyClub/circle-ci-conditional-job-test/blob/new-new-stroud/.circleci/config.yml#L23-L25)

**Comment by symroe** [comment]
> doesn't stop the workflow, it stops the step cleanly

Maybe think of it like and early return of a method rather than a `sys.exit()`?

**Comment by GeoWill** [comment]
OK, I think I've got it now. Clearly wishful thinking as I convinced myself you'd solved that with the `step halt` approach!

**Pull request merged**